sklearn/utils/_array_out.py

```diff
@@ -0,0 +1,122 @@
+from functools import wraps
+from functools import partial
+from inspect import signature
+
+import scipy.sparse as sp_sparse
+
+
+def _get_feature_names(X):
+    """Get feature names of a dataframe."""
+    if hasattr(X, "iloc"):  # duck-type dataframe
+        return getattr(X, "columns", None)
+
+
+def _get_index(X):
+    """Get the index of a dataframe."""
+    if hasattr(X, "iloc"):  # duck-type dataframe
+        return getattr(X, "index", None)
+
+
+def _check_get_feature_names_out(get_feature_names_out, estimator,
+                                 n_feature_out):
+    """Check and convert get_feature_names_out into a callable."""
+    if callable(get_feature_names_out):
+        get_feature_names_out_callable = get_feature_names_out(estimator)
+    elif get_feature_names_out == 'one_to_one':
+        def get_feature_names_out_callable(names):
+            return names
+    else:
+        # get_feature_names_out == 'class_name'
+        class_name = estimator.__class__.__name__.lower()
+
+        def get_feature_names_out_callable():
+            return [f"{class_name}{i}" for i in range(n_feature_out)]
+
+    # feature names in can have zero or one argument. For one argument
+    # it would be the input feature names
+    parameters = signature(get_feature_names_out_callable).parameters
+
+    if parameters:
+        feature_names_in = getattr(estimator, "feature_names_in_", None)
+        get_feature_names_out_callable = partial(
+            get_feature_names_out_callable, feature_names_in)
+    return get_feature_names_out_callable
+
+
+def _make_array_out(X_out, index, get_feature_names_out, *,
```

Comment:
```suggestion
def _make_array_out(X_out, *, index, get_feature_names_out,
```

Reply:
also, wouldn't it make more sense for this function to accept the `feature_names_out` instead of `get_feature names_out`?

Reply:
I guess not, because of the way it's used in the decorator?